<a href="https://colab.research.google.com/github/imugagga/Sales-Predictions---JFood-Sores/blob/main/Sales_Predictions_Mugagga_Innocent.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Sales-Predictions---JFood-Sores**

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Week 1/Uploaded Data/sales_predictions.csv')
df.head(3)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDA15         9.30          Low Fat         0.016047        Dairy   
1           DRC01         5.92          Regular         0.019278  Soft Drinks   
2           FDN15        17.50          Low Fat         0.016760         Meat   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  249.8092            OUT049                       1999      Medium   
1   48.2692            OUT018                       2009      Medium   
2  141.6180            OUT049                       1999      Medium   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0               Tier 1  Supermarket Type1          3735.1380  
1               Tier 3  Supermarket Type2           443.4228  
2               Tier 1  Supermarket Type1          2097.2700

In [3]:
df.tail(3)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
8520           NCJ29        10.60          Low Fat         0.035186   
8521           FDN46         7.21          Regular         0.145221   
8522           DRG01        14.80          Low Fat         0.044878   

               Item_Type  Item_MRP Outlet_Identifier  \
8520  Health and Hygiene   85.1224            OUT035   
8521         Snack Foods  103.1332            OUT018   
8522         Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700

*Data Inspection*

In [4]:
# No. rows and columns?
df.shape

(8523, 12)

In [5]:
len(df)

8523

In [6]:
# Datatypes of each variable
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [7]:
# Checking for duplicates
df.duplicated().sum()

0

In [8]:
#View Columns with missing Values
df.isnull().sum()   # 2 columns with missing values

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [9]:
#Inspect the number of unique values in each of the columns 
df.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [10]:
#Inspect the number of unique values in each of the columns with missing data.
df[['Item_Weight','Outlet_Size']].nunique()

Item_Weight    415
Outlet_Size      3
dtype: int64

In [11]:
#Impect the columns with the missing values
df['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

*Address the missing values and do it! (This requires your judgement, so explain your choice).*

In [12]:
#For the Column 'Outlet_Size'

#View the percentage of missing data 
#Count the Non null Value
Non_nullcells = df['Outlet_Size'].count()

#Obtain the length of the column
Totalcells = len(df['Outlet_Size'])
#Calculate the percentage of the missing data
Empty_cells = Totalcells - Non_nullcells
Percent = ((Empty_cells/Totalcells)*100).round(2)
print(f'The empty cells are {Empty_cells} and they are {Percent}% of the data in the column')

The empty cells are 2410 and they are 28.28% of the data in the column


In [13]:
#Relating the Column with the 'Outlet_size' column

df.groupby(['Outlet_Size'])['Outlet_Type'].value_counts()

Outlet_Size  Outlet_Type      
High         Supermarket Type1     932
Medium       Supermarket Type3     935
             Supermarket Type1     930
             Supermarket Type2     928
Small        Supermarket Type1    1860
             Grocery Store         528
Name: Outlet_Type, dtype: int64

***Explanation for Outlet_Size:***

From the observation above the most supermarkets are of the medium size and thus the missing values in this column shall be filled with the medium size  obtained using the mode function so as not to loose the data in this column.

1. The Missing data is a small percentage ie 28.28 % of the data in the column 
2. Droping the column would mean loosing the entire column attribute of the data
3. Dropping the rows would mean loosing all sale details from 2410 entries from the data set. 

In [14]:
#For the Column 'Item_Weight'

#View the percentage of missing data 
#Count the Non null Value
Non_nullcells1 = df['Item_Weight'].count()

#Obtain the length of the column
Totalcells1 = len(df['Item_Weight'])
#Calculate the percentage of the missing data
Empty_cells1 = Totalcells1 - Non_nullcells1
Percent1 = ((Empty_cells1/Totalcells1)*100).round(2)
print(f'The empty cells are {Empty_cells1} and they are {Percent1}% of the data in the column')

The empty cells are 1463 and they are 17.17% of the data in the column


***Explanation for Item_Weight:***

Since the Weight of the items purchased is numeriacal and it's a very important attribute, it will help to tell consumption patterns and further analysis in future , i dont want to drop the column thus i have chosen to replace the missing values with the median of the column data.

1. The Missing data is a small percentage ie 17.17 % of the data in the column
2. Droping the column would mean loosing the entire column attribute of the data
3. Dropping the rows would mean loosing all sale details from 1463 entries from the data set.

In [15]:
#Obtaining the mode of the column 'Outlet_Size'.
Size_mode = df['Outlet_Size'].mode()[0]

#Filling empty cells
df['Outlet_Size'].fillna(Size_mode,inplace=True)

print('Mode used is: ',Size_mode)

Mode used is:  Medium


In [16]:
#Obtaining the Median of the column 'Item_Weight'.
Weight_median = df['Item_Weight'].median()

#Filling empty cells
df['Item_Weight'].fillna(Weight_median,inplace=True)

print('Median used is: ',Weight_median)

Median used is:  12.6


In [17]:
#Inspecting the cleaned dataframe
df.isnull().sum()
#From this the outpuit confirms that all missing values have been filled up.

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

***Checking for any inconsistances in the Data***

In [18]:
#Selecting columns with object data type
df.select_dtypes(include=['object']).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Item_Identifier       8523 non-null   object
 1   Item_Fat_Content      8523 non-null   object
 2   Item_Type             8523 non-null   object
 3   Outlet_Identifier     8523 non-null   object
 4   Outlet_Size           8523 non-null   object
 5   Outlet_Location_Type  8523 non-null   object
 6   Outlet_Type           8523 non-null   object
dtypes: object(7)
memory usage: 466.2+ KB


In [19]:
df.select_dtypes(include=['object']).nunique()

Item_Identifier         1559
Item_Fat_Content           5
Item_Type                 16
Outlet_Identifier         10
Outlet_Size                3
Outlet_Location_Type       3
Outlet_Type                4
dtype: int64

In [20]:
#Identifying the inconsistent data in the column
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [21]:
#Fixing the inconsistent data using the replace function.
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace(['LF','low fat'],'Low Fat').replace('reg','Regular')

In [22]:
#Confirming that the  column data is consistent
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [23]:
#Obtain the summary statistics for numerical columns.
df.describe().round(3)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
count     8523.000         8523.000  8523.000                   8523.000   
mean        12.813            0.066   140.993                   1997.832   
std          4.227            0.052    62.275                      8.372   
min          4.555            0.000    31.290                   1985.000   
25%          9.310            0.027    93.826                   1987.000   
50%         12.600            0.054   143.013                   1999.000   
75%         16.000            0.095   185.644                   2004.000   
max         21.350            0.328   266.888                   2009.000   

       Item_Outlet_Sales  
count           8523.000  
mean            2181.289  
std             1706.500  
min               33.290  
25%              834.247  
50%             1794.331  
75%             3101.296  
max            13086.965

In [24]:
#View a sample of the Cleaned Dataframe
df.sample(5)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7413           FDH53       20.500          Regular         0.019278   
1361           FDX36        9.695          Regular         0.129009   
4642           DRG15        6.130          Low Fat         0.077170   
2234           FDL43       10.100          Low Fat         0.027217   
4990           FDE46       18.600          Low Fat         0.015859   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
7413  Frozen Foods   82.8592            OUT018                       2009   
1361  Baking Goods  226.9404            OUT017                       2007   
4642         Dairy   63.0536            OUT017                       2007   
2234          Meat   76.0670            OUT017                       2007   
4990   Snack Foods  152.9366            OUT017                       2007   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
7413      Medium               Tier 3  Supermarket Type2          1733.7432  
1361      Medium               Tier 2  Supermarket Type1          2700.4848  
4642      Medium               Tier 2  Supermarket Type1          1225.0720  
2234      Medium               Tier 2  Supermarket Type1          1454.7730  
4990      Medium               Tier 2  Supermarket Type1           453.4098